### PR TITLE
Improvements to entity list loading times

### PR DESCRIPTION
**Pull request by huffman**
Do not review yet, I need to give it a clean up pass

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12682/HighFidelity-Beta-PR13943-335f66d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12682/HighFidelity-Beta-PR13943-335f66d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12682/HighFidelity-Beta-Interface-PR13943-335f66d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12682/HighFidelity-Beta-PR13943-335f66d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12682/HighFidelity-Beta-Interface-PR13943-335f66d.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12776/HighFidelity-Beta-PR13943-574ff53.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12776/HighFidelity-Beta-PR13943-574ff53.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12776/HighFidelity-Beta-Interface-PR13943-574ff53.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12776/HighFidelity-Beta-PR13943-574ff53.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12776/HighFidelity-Beta-Interface-PR13943-574ff53.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12787/HighFidelity-Beta-PR13943-9b9785d.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12789/HighFidelity-Beta-PR13943-3cc90d3.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12789/HighFidelity-Beta-PR13943-3cc90d3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12789/HighFidelity-Beta-Interface-PR13943-3cc90d3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12789/HighFidelity-Beta-PR13943-3cc90d3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12789/HighFidelity-Beta-Interface-PR13943-3cc90d3.exe

**Comment by huffman** [comment]
# Test Plan

 * https://github.com/highfidelity/hifi/pull/13943

**Comment by huffman** [comment]
Cheyenne is looking at this currently

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12819/HighFidelity-Beta-PR13943-0fadfd2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12819/HighFidelity-Beta-Interface-PR13943-0fadfd2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12819/HighFidelity-Beta-PR13943-0fadfd2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12819/HighFidelity-Beta-Interface-PR13943-0fadfd2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12822/HighFidelity-Beta-PR13943-5aa5552.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12822/HighFidelity-Beta-Interface-PR13943-5aa5552.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12822/HighFidelity-Beta-PR13943-5aa5552.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12822/HighFidelity-Beta-Interface-PR13943-5aa5552.exe

**Comment by QliCknox** [comment]
Currently Testing

**Comment by QliCknox** [comment]
Pass @conklin94122 

Entity list loads quickly upon loading into a domain.

File: scripts/system/libraries/entityList.js [hunk 1]
```diff
@@ -116,59 +137,66 @@ EntityListTool = function(shouldUseEditTabletApp) {
...
 
-        var ids;
-        if (filterInView) {
-            ids = Entities.findEntitiesInFrustum(Camera.frustum);
-        } else {
-            ids = Entities.findEntities(MyAvatar.position, searchRadius);
-        }
+            var ids;
+            PROFILE("findEntities", function() {
+            if (filterInView) {
```
Comment (dback2) [inline scripts/system/libraries/entityList.js]: minor thing but indentations off here

File: scripts/system/libraries/entityList.js [hunk 2]
```diff
@@ -116,59 +137,66 @@ EntityListTool = function(shouldUseEditTabletApp) {
...
+                ids = Entities.findEntitiesInFrustum(Camera.frustum);
+            } else {
+                ids = Entities.findEntities(MyAvatar.position, searchRadius);
+            }
+            });
+
+            var cameraPosition = Camera.position;
+            PROFILE("getProperties", function() {
+            for (var i = 0; i < ids.length; i++) {
```
Comment (dback2) [inline scripts/system/libraries/entityList.js]: and here

File: scripts/system/html/js/entityList.js [hunk 3]
```diff
@@ -363,31 +480,20 @@ function loaded() {
...
-                                      newEntities[i].locked ? LOCKED_GLYPH : null,
-                                      newEntities[i].visible ? VISIBLE_GLYPH : null,
-                                      newEntities[i].verticesCount, newEntities[i].texturesCount, newEntities[i].texturesSize,
-                                      newEntities[i].hasTransparent ? TRANSPARENCY_GLYPH : null,
-                                      newEntities[i].isBaked ? BAKED_GLYPH : null,
-                                      newEntities[i].drawCalls,
-                                      newEntities[i].hasScript ? SCRIPT_GLYPH : null);
+                    PROFILE("update", function() {
+                        var newEntities = data.entities;
+                        if (newEntities && newEntities.length == 0) {
```
Comment (dback2) [inline scripts/system/html/js/entityList.js]: ===

File: scripts/system/html/js/entityList.js [hunk 4]
```diff
@@ -89,33 +122,34 @@ function loaded() {
...
-                for (var entity in entityList.visibleItems) {
-                    if (clickedItemFound === -1 && this.dataset.entityId == entityList.visibleItems[entity].values().id) {
-                        clickedItemFound = entity;
-                    } else if(previousItemFound === -1 && selectedEntities[0] == entityList.visibleItems[entity].values().id) {
-                        previousItemFound = entity;
+                let previousItemFound = -1;
+                let clickedItemFound = -1;
+                for (let i = 0, len = visibleEntities.length; i < len; ++i) {
+                    let entity = visibleEntities[i];
+                    if (clickedItemFound === -1 && entityID == entity.id) {
```
Comment (dback2) [inline scripts/system/html/js/entityList.js]: ===

File: scripts/system/html/js/entityList.js [hunk 5]
```diff
@@ -156,64 +196,144 @@ function loaded() {
...
+        function getFilename(url) {
+            let urlParts = url.split('/');
+            return urlParts[urlParts.length - 1];
+        }
 
-            var urlParts = url.split('/');
-            var filename = urlParts[urlParts.length - 1];
+        elFilter.onkeyup = refreshEntityList;
+        elFilter.onpaste = refreshEntityList;
+        elFilter.onchange = refreshEntityList;
```
Comment (dback2) [inline scripts/system/html/js/entityList.js]: Does this not get overwritten by the onchange set on line 543?

File: scripts/system/html/js/entityList.js [hunk 6]
```diff
@@ -156,64 +196,144 @@ function loaded() {
...
+                PROFILE("update-dom", function() {
+                    elEntityTableBody.innerHTML = '';
+                    for (let i = 0, len = visibleEntities.length; i < len; ++i) {
+                        elEntityTableBody.append(visibleEntities[i].el);
+                    }
+                });
+            });
         }
 
-        function scheduleRefreshEntityList() {
```
[the remark this answers is not in the document]
Comment (huffman) [reply scripts/system/html/js/entityList.js]: No, it was specifically for listjs which required a separate sort/filter with each change. We do this for every update to the list now. We might want to add some buffering in the future, but in my testing it was not needed.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12838/HighFidelity-Beta-PR13943-4b8f5ea.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12838/HighFidelity-Beta-PR13943-4b8f5ea.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12838/HighFidelity-Beta-Interface-PR13943-4b8f5ea.exe

**Pull request merged**